dtmsvr/utils.go

```diff
@@ -50,3 +54,82 @@ func GetTransGlobal(gid string) *TransGlobal {
 	//nolint:staticcheck
 	return &TransGlobal{TransGlobalStore: *trans}
 }
+
+type iface struct {
+	itab, data uintptr
+}
+
+type valueCtx struct {
+	context.Context
+	key, value any
+}
+
+type cancelCtx struct {
+	context.Context
+}
+
+type timerCtx struct {
+	cancelCtx *cancelCtx
+}
+
+func (*timerCtx) Deadline() (deadline time.Time, ok bool) {
+	return
+}
+
+func (*timerCtx) Done() <-chan struct{} {
+	return nil
+}
+
+func (*timerCtx) Err() error {
+	return nil
+}
+
+func (*timerCtx) Value(key any) any {
+	return nil
+}
+
+func (e *timerCtx) String() string {
+	return ""
+}
+
+// CopyContext copy context with value and grpc metadata
+// if raw context is nil, return nil
+func CopyContext(ctx context.Context) context.Context {
```

Comment:
Is there an elegant way to copy context?

Reply:
I will make an attempt to optimize it.

Reply:
The implementation involving reflection and the unsafe package does indeed raise concerns. In the Go version 1.19.3 used in the CI environment, the structure of timerCtx appears as follows:

![image](https://github.com/dtm-labs/dtm/assets/75628309/4eefca81-c067-4090-aaf6-486a915f76bb)

However, in the Go version 1.20.5 on my local setup, the structure of timerCtx is different, displaying a combination-inheritance composition:

![image](https://github.com/dtm-labs/dtm/assets/75628309/bd10fb6c-c712-4686-9de9-c8844f392b67)

In my implementation, I utilize the "unsafe" package to convert a uintptr to a specific timerCtx type. This results in incompatibility between different versions of Go, which can cause issues with the dtm functionality. Do you have any suggestions regarding this?

Reply:
> Is there an elegant way to copy context?
> 
> I remember that, all values can be overridden in context, so I guess we may pass in another timeout to override the default

No, if the current deadline is already earlier than the new one, no action will be taken.
![image](https://github.com/dtm-labs/dtm/assets/75628309/a58f2b96-e168-425b-9c2f-5d274a272969)

Reply:
Should not rely on the internal implementation of Context. My suggestion:
```
type asyncCtx struct {
Context
}

func (*emptyCtx) Deadline() (deadline time.Time, ok bool) {
	return
}

func (*emptyCtx) Done() <-chan struct{} {
	return nil
}

func newAysncCtx(parent Context) asyncCtx {
  return asyncCtx{Context: context}
}
```

Reply:
why not use anonymous member, so that you do not need to implement Value and Err?